File: QRLoginManager.cs
// QRLoginManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using QRCoder;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;

namespace QRLoginManager
{
    class Program
    {
        private const int SaltSize = 16;
        private const int NonceSize = 12;
        private const int KeySize = 32;
        private const int Iterations = 100000;

        static byte[] DeriveKey(string password, byte[] salt)
        {
            using (var derive = new Rfc2898DeriveBytes(password, salt, Iterations, HashAlgorithmName.SHA256))
            {
                return derive.GetBytes(KeySize);
            }
        }

        static byte[] EncryptData(byte[] data, string password)
        {
            using (var aes = new AesGcm())
            {
                var salt = new byte[SaltSize];
                var nonce = new byte[NonceSize];
                using (var rng = RandomNumberGenerator.Create())
                {
                    rng.GetBytes(salt);
                    rng.GetBytes(nonce);
                }
                var key = DeriveKey(password, salt);
                aes.Key = key;
                aes.Nonce = nonce;
                var ciphertext = new byte[data.Length];
                var tag = new byte[16];
                aes.Encrypt(data, ciphertext, tag);
                var combined = new byte[salt.Length + nonce.Length + ciphertext.Length + tag.Length];
                Buffer.BlockCopy(salt, 0, combined, 0, salt.Length);
                Buffer.BlockCopy(nonce, 0, combined, salt.Length, nonce.Length);
                Buffer.BlockCopy(ciphertext, 0, combined, salt.Length + nonce.Length, ciphertext.Length);
                Buffer.BlockCopy(tag, 0, combined, salt.Length + nonce.Length + ciphertext.Length, tag.Length);
                return combined;
            }
        }

        static byte[] DecryptData(byte[] encrypted, string password)
        {
            if (encrypted.Length < SaltSize + NonceSize + 16) throw new Exception("Invalid data");
            var salt = new byte[SaltSize];
            var nonce = new byte[NonceSize];
            var tag = new byte[16];
            Buffer.BlockCopy(encrypted, 0, salt, 0, salt.Length);
            Buffer.BlockCopy(encrypted, salt.Length, nonce, 0, nonce.Length);
            int cipherLen = encrypted.Length - salt.Length - nonce.Length - tag.Length;
            var ciphertext = new byte[cipherLen];
            Buffer.BlockCopy(encrypted, salt.Length + nonce.Length, ciphertext, 0, cipherLen);
            Buffer.BlockCopy(encrypted, salt.Length + nonce.Length + cipherLen, tag, 0, tag.Length);
            var key = DeriveKey(password, salt);
            using (var aes = new AesGcm())
            {
                aes.Key = key;
                aes.Nonce = nonce;
                var plaintext = new byte[cipherLen];
                aes.Decrypt(ciphertext, plaintext, tag);
                return plaintext;
            }
        }

        static List<string> GenerateCodes(int count, int length)
        {
            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
            var rng = new Random();
            var codes = new List<string>();
            for (int i = 0; i < count; i++)
            {
                var sb = new StringBuilder();
                for (int j = 0; j < length; j++)
                    sb.Append(chars[rng.Next(chars.Length)]);
                codes.Add(sb.ToString());
            }
            return codes;
        }

        class BackupData
        {
            public string Created { get; set; }
            public List<string> Codes { get; set; }
        }

        static void SaveToFile(BackupData data, string password, string filename)
        {
            var json = JsonSerializer.SerializeToUtf8Bytes(data);
            var encrypted = EncryptData(json, password);
            File.WriteAllBytes(filename, encrypted);
            Console.WriteLine($"Saved encrypted backup to {filename}");
        }

        static BackupData LoadFromFile(string filename, string password)
        {
            var encrypted = File.ReadAllBytes(filename);
            var decrypted = DecryptData(encrypted, password);
            return JsonSerializer.Deserialize<BackupData>(decrypted);
        }

        static void ListCodes(string filename, string password)
        {
            var data = LoadFromFile(filename, password);
            Console.WriteLine($"Backup: {filename}");
            Console.WriteLine($"Created: {data.Created}");
            Console.WriteLine("Codes:");
            foreach (var code in data.Codes)
                Console.WriteLine($"  {code}");
        }

        static void GenerateQR(string filename, string password, string output)
        {
            var data = LoadFromFile(filename, password);
            var codesStr = string.Join("\n", data.Codes);
            using (var qrGenerator = new QRCodeGenerator())
            {
                var qrData = qrGenerator.CreateQrCode(codesStr, QRCodeGenerator.ECCLevel.Q);
                var qrCode = new QRCode(qrData);
                using (var qrImage = qrCode.GetGraphic(20))
                {
                    qrImage.Save(output, System.Drawing.Imaging.ImageFormat.Png);
                }
            }
            Console.WriteLine($"QR code saved to {output}");
        }

        static void CheckCode(string filename, string password, string code)
        {
            var data = LoadFromFile(filename, password);
            if (data.Codes.Contains(code))
                Console.WriteLine($"Code '{code}' is valid (found in backup).");
            else
                Console.WriteLine($"Code '{code}' is NOT in backup.");
        }

        static void Main(string[] args)
        {
            var dict = new Dictionary<string, string>();
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i].StartsWith("--"))
                {
                    var key = args[i].Substring(2);
                    if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                        dict[key] = args[++i];
                    else
                        dict[key] = "";
                }
            }

            try
            {
                if (dict.ContainsKey("generate"))
                {
                    int count = int.Parse(dict["generate"]);
                    string output = dict.GetValueOrDefault("output");
                    string password = dict.GetValueOrDefault("password");
                    if (string.IsNullOrEmpty(output) || string.IsNullOrEmpty(password))
                        throw new Exception("--output and --password required");
                    var data = new BackupData { Created = DateTime.UtcNow.ToString("o"), Codes = GenerateCodes(count, 8) };
                    SaveToFile(data, password, output);
                }
                else if (dict.ContainsKey("list"))
                {
                    string file = dict["list"];
                    string password = dict.GetValueOrDefault("password");
                    if (string.IsNullOrEmpty(password))
                        throw new Exception("--password required");
                    ListCodes(file, password);
                }
                else if (dict.ContainsKey("qr"))
                {
                    string file = dict["qr"];
                    string password = dict.GetValueOrDefault("password");
                    string output = dict.GetValueOrDefault("output");
                    if (string.IsNullOrEmpty(password) || string.IsNullOrEmpty(output))
                        throw new Exception("--password and --output required for QR");
                    GenerateQR(file, password, output);
                }
                else if (dict.ContainsKey("check"))
                {
                    string code = dict["check"];
                    string file = dict.GetValueOrDefault("list");
                    string password = dict.GetValueOrDefault("password");
                    if (string.IsNullOrEmpty(file) || string.IsNullOrEmpty(password))
                        throw new Exception("--list and --password required for check");
                    CheckCode(file, password, code);
                }
                else
                {
                    Console.WriteLine("Use --help for usage.");
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: qr_login_manager.rs
// qr_login_manager.rs
use clap::{App, Arg};
use aes_gcm::{Aes256Gcm, Key, Nonce};
use aes_gcm::aead::{Aead, NewAead};
use rand::RngCore;
use serde::{Deserialize, Serialize};
use serde_json;
use std::fs;
use std::io::Write;
use std::time::{SystemTime, UNIX_EPOCH};
use qrcode::QrCode;
use image::Luma;

const SALT_LEN: usize = 16;
const NONCE_LEN: usize = 12;
const KEY_LEN: usize = 32;
const ITERATIONS: u32 = 100_000;

fn derive_key(password: &str, salt: &[u8]) -> [u8; KEY_LEN] {
    use ring::pbkdf2;
    let mut key = [0u8; KEY_LEN];
    pbkdf2::derive(
        pbkdf2::PBKDF2_HMAC_SHA256,
        std::num::NonZeroU32::new(ITERATIONS).unwrap(),
        salt,
        password.as_bytes(),
        &mut key,
    );
    key
}

fn encrypt_data(data: &[u8], password: &str) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut salt = [0u8; SALT_LEN];
    let mut nonce = [0u8; NONCE_LEN];
    let mut rng = rand::thread_rng();
    rng.fill_bytes(&mut salt);
    rng.fill_bytes(&mut nonce);
    let key = derive_key(password, &salt);
    let cipher = Aes256Gcm::new(Key::from_slice(&key));
    let ciphertext = cipher.encrypt(Nonce::from_slice(&nonce), data)?;
    let mut result = Vec::with_capacity(SALT_LEN + NONCE_LEN + ciphertext.len());
    result.extend_from_slice(&salt);
    result.extend_from_slice(&nonce);
    result.extend_from_slice(&ciphertext);
    Ok(result)
}

fn decrypt_data(encrypted: &[u8], password: &str) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    if encrypted.len() < SALT_LEN + NONCE_LEN {
        return Err("Invalid encrypted data".into());
    }
    let salt = &encrypted[..SALT_LEN];
    let nonce = &encrypted[SALT_LEN..SALT_LEN + NONCE_LEN];
    let ciphertext = &encrypted[SALT_LEN + NONCE_LEN..];
    let key = derive_key(password, salt);
    let cipher = Aes256Gcm::new(Key::from_slice(&key));
    let plaintext = cipher.decrypt(Nonce::from_slice(nonce), ciphertext)?;
    Ok(plaintext)
}

#[derive(Serialize, Deserialize)]
struct BackupData {
    created: String,
    codes: Vec<String>,
}

fn generate_codes(count: usize, length: usize) -> Vec<String> {
    const CHARS: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    let mut rng = rand::thread_rng();
    let mut codes = Vec::with_capacity(count);
    for _ in 0..count {
        let mut code = String::with_capacity(length);
        for _ in 0..length {
            let idx = rng.next_u32() as usize % CHARS.len();
            code.push(CHARS[idx] as char);
        }
        codes.push(code);
    }
    codes
}

fn save_to_file(data: &BackupData, password: &str, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    let json = serde_json::to_vec(data)?;
    let enc = encrypt_data(&json, password)?;
    fs::write(filename, enc)?;
    println!("Saved encrypted backup to {}", filename);
    Ok(())
}

fn load_from_file(filename: &str, password: &str) -> Result<BackupData, Box<dyn std::error::Error>> {
    let enc = fs::read(filename)?;
    let plain = decrypt_data(&enc, password)?;
    let data: BackupData = serde_json::from_slice(&plain)?;
    Ok(data)
}

fn list_codes(filename: &str, password: &str) -> Result<(), Box<dyn std::error::Error>> {
    let data = load_from_file(filename, password)?;
    println!("Backup: {}", filename);
    println!("Created: {}", data.created);
    println!("Codes:");
    for code in data.codes {
        println!("  {}", code);
    }
    Ok(())
}

fn generate_qr(filename: &str, password: &str, output: &str) -> Result<(), Box<dyn std::error::Error>> {
    let data = load_from_file(filename, password)?;
    let codes_str = data.codes.join("\n");
    let qr = QrCode::new(codes_str)?;
    let img = qr.render::<Luma<u8>>().build();
    img.save(output)?;
    println!("QR code saved to {}", output);
    Ok(())
}

fn check_code(filename: &str, password: &str, code: &str) -> Result<(), Box<dyn std::error::Error>> {
    let data = load_from_file(filename, password)?;
    if data.codes.contains(&code.to_string()) {
        println!("Code '{}' is valid (found in backup).", code);
    } else {
        println!("Code '{}' is NOT in backup.", code);
    }
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = App::new("QR Login Manager")
        .arg(Arg::with_name("generate").long("generate").takes_value(true).help("Generate N codes"))
        .arg(Arg::with_name("output").long("output").takes_value(true).help("Output file"))
        .arg(Arg::with_name("password").long("password").takes_value(true).help("Password"))
        .arg(Arg::with_name("list").long("list").takes_value(true).help("List codes from file"))
        .arg(Arg::with_name("qr").long("qr").takes_value(true).help("Generate QR from file"))
        .arg(Arg::with_name("check").long("check").takes_value(true).help("Check code in backup"))
        .get_matches();

    if let Some(count_str) = matches.value_of("generate") {
        let count = count_str.parse()?;
        let output = matches.value_of("output").ok_or("--output required")?;
        let password = matches.value_of("password").ok_or("--password required")?;
        let codes = generate_codes(count, 8);
        let data = BackupData {
            created: format!("{:?}", SystemTime::now().duration_since(UNIX_EPOCH).unwrap()),
            codes,
        };
        save_to_file(&data, password, output)?;
    } else if let Some(file) = matches.value_of("list") {
        let password = matches.value_of("password").ok_or("--password required")?;
        list_codes(file, password)?;
    } else if let Some(file) = matches.value_of("qr") {
        let password = matches.value_of("password").ok_or("--password required")?;
        let output = matches.value_of("output").ok_or("--output required for QR")?;
        generate_qr(file, password, output)?;
    } else if let Some(code) = matches.value_of("check") {
        let file = matches.value_of("list").ok_or("--list required for check")?;
        let password = matches.value_of("password").ok_or("--password required")?;
        check_code(file, password, code)?;
    } else {
        println!("Use --help for usage.");
    }
    Ok(())
}
